Refactor server-sent events parsing

valadaptive opened the pull request:
Previously, there were 4 slightly different server-sent event parsers for the 4 different streaming API implementations. I've replaced them all with one server-sent event transform stream, which takes in a `Uint8Array` stream and turns it into a `MessageEvent` stream. It implements the entire [server-sent event stream spec](https://html.spec.whatwg.org/multipage/server-sent-events.html#parsing-an-event-stream), so the AI frontends don't need to implement their own bespoke parsers.

I've actually released this code [as an npm package](https://www.npmjs.com/package/@server-sent-stream/web), but I've vendored it in case it needs to be changed. I did create a [test suite](https://github.com/valadaptive/server-sent-stream/blob/6e5f2b66a74bd1a66ee01bef2e86ba807f7b0512/test/sse-stream.test.ts) for it, though, so hopefully the implementation's solid.

valadaptive committed:
Refactor server-sent events parsing

Create one server-sent events stream class which implements the entire
spec (different line endings, chunking, etc) and use it in all the
streaming generators.

Cohee1207 commented:
Can you confirm the following?
1. It works when events are split by a single newline instead of two newlines as per spec (for legacy servers).
2. It works when a client receives a partial/incomplete message chunk (through Cloudflare gate, etc).
3. If the client receives a JSON body instead of an SSE stream, the body is correctly parsed and thrown as an error.

valadaptive commented:
1. Which servers do this? That seems like it's so against the spec that it would break any servers that *do* comply with the spec.
2. Yes; I've tested it and confirmed it works with partial message fragments split up in a variety of weird places.
3. Which APIs sometimes return JSON and sometimes return SSE streams from the same endpoint? I might need to do some testing to figure out how to handle that.

valadaptive committed:
Don't stream events if the API returned a 4xx code

valadaptive commented:
I updated the code to treat the response as JSON instead of an event stream if the server returns a 4xx error, which I think handles point 3. Still not sure when point 1 occurs.

Cohee1207 commented:
1. Some old proxies, I don't have an exact list, but that was the intention behind the "legacy streaming" switch.
3. NovelAI / Aphrodite engine to name a few. tryParseStreamingError was the function(s) that handled it for them.

valadaptive commented:
I'm a bit confused - what are they proxying? ST's server already handles CORS stuff; what do the proxies do?

Cohee1207 commented:
Reverse proxies for OpenAI and Anthropic. Specifically this project and similar.
https://gitgud.io/khanon/oai-reverse-proxy

We don't provide official support for users of such methods, but I'd like to keep it functional still even if some of them may implement streaming incorrectly and still in use by some.

valadaptive commented:
It looks like oai-reverse-proxy, at least the latest version, [handles SSE properly](https://gitgud.io/khanon/oai-reverse-proxy/-/blob/main/src/proxy/middleware/response/streaming/sse-stream-adapter.ts?ref_type=heads). I can add back support for legacy streaming if it's really necessary, but from looking in the Discord it seems like most people are confused about its functionality and/or enable it when it's unnecessary or actively counterproductive.

Are there multiple proxies that implement SSE incorrectly, or was it just one old one? From a skim over the Discord, I can't find anyone who actually uses the feature.

valadaptive commented:
Every single reverse proxy I can find seems to return events with two newlines in between:
- [PawanOsman/ChatGPT implements SSE correctly](https://github.com/PawanOsman/ChatGPT/blob/98a35b53af78a5e631e9b67bbaf2e46644d5a309/routes.js#L63)
- [khanon/oai-reverse-proxy implements SSE correctly](https://gitgud.io/khanon/oai-reverse-proxy/-/blob/94d4efe9bb0dc160447e1354180e7273920b4c70/src/proxy/middleware/response/handle-streamed-response.ts#L75)
- [juzeon/poe-openai-proxy implements SSE correctly](https://github.com/juzeon/poe-openai-proxy/blob/2e5b13683d44ef194376bb08769c6b8f47f29aa0/router/router.go#L106)
- [KenyonY/openai-forward implements SSE correctly](https://github.com/KenyonY/openai-forward/blob/94df78321e2642f91471116d5eabca7e5fd88a55/openai_forward/cache/chat_completions.py#L254)

There are a bunch more that transparently pass the stream straight through, or wait for the entire stream then resend it (which is not a very good streaming API but complies with the spec).

I can't find any proxy that separates data chunks with a single newline, and the commit history doesn't go back far enough to find out why legacy streaming mode was implemented in the first place.

Cohee1207 commented:
This was added after this PR specifically, which reportedly broke streaming for some proxies at the time. https://github.com/SillyTavern/SillyTavern/pull/340
I can assume that 6 months were enough for every legacy proxy maintainer to update their streaming implementation. In any case, after this PR legacy streaming toggle will do absolutely nothing and should be dropped.

valadaptive commented:
Yup, https://github.com/PandarusAnon/slaude/pull/9 fixed that 7 months ago, so I'll remove the legacy streaming checkbox, and we can mention to Slaude users specifically that they might need to update Slaude in the next release notes.

valadaptive committed:
Remove legacy_streaming setting

This was a workaround for older versions of Slaude that implemented SSE
improperly. This was fixed in Slaude 7 months ago, so the workaround can
be removed.

valadaptive committed:
Properly forward status codes from streams

Cohee1207 committed:
Fix streaming processor error handler hooks

valadaptive committed:
Remove ignoreBOM

Apparently the ignoreBOM option actually means "include the BOM". I've
added a test for this in my own repository, and will also be submitting
a pull request to MDN to clarify this in their documentation.

valadaptive committed:
Refactor event stream parsing

I was really overcomplicating this before; this is simpler and faster.
Passes my existing test suite.

valadaptive committed:
Refactor error handling

Remove the StreamingProcessor.hook method and use a try-catch block to
await the generator promise and set the generator, handling errors with
onError if it fails.

valadaptive committed:
Default event type to 'message' if it's empty

Added to the upstream test suite. Matches the spec.

Cohee1207 committed:
Avoid sending 401 in forwarded responses.
Status code 401 resets the client Basic auth, so we replace it with 400. This can produce an interesting artifact as "400 Unauthorized", but it's not out of spec.

"The reason phrases listed here are only recommendations -- they can be replaced by local equivalents or left out altogether without affecting the protocol."

https://www.rfc-editor.org/rfc/rfc9110.html#name-overview-of-status-codes

Cohee1207 commented:
Tested, appears to be working on all sources that support streaming, and correctly showing/parsing errors.
